src/geometry/manifold/algebra/left_invariant_derivation.lean

```diff
@@ -0,0 +1,195 @@
+/-
+Copyright © 2020 Nicolò Cavalleri. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Nicolò Cavalleri
+-/
+
+import geometry.manifold.derivation_bundle
+
+/-!
+
+# Left invariant derivations
+
+In this file we define the concept of left invariant derivation for a Lie group. The concept is
+analogous to the more classical concept of left invariant vector fields, and it holds that the
+derivation associated to a vector field is left invariant iff the field is.
+
+Moreover we prove that `left_invariant_derivation I G` has the structure of a Lie algebra, hence
+implementing one of the possible definitions of the Lie algebra attached to a Lie group.
+
+-/
+
+noncomputable theory
+
+open_locale lie_group manifold derivation
+
+variables {𝕜 : Type*} [nondiscrete_normed_field 𝕜]
+{E : Type*} [normed_group E] [normed_space 𝕜 E]
+{H : Type*} [topological_space H] (I : model_with_corners 𝕜 E H)
+(G : Type*) [topological_space G] [charted_space H G] [monoid G] [has_smooth_mul I G] (g h : G)
+
+-- Generate trivial has_sizeof instance. It prevents weird type class inference timeout problems
+local attribute [nolint instance_priority, instance, priority 10000]
+private def disable_has_sizeof {α} : has_sizeof α := ⟨λ _, 0⟩
+
+/--
+Left-invariant global derivations.
+
+A global derivation is left-invariant if it is equal to its pullback along left multiplication by
+an arbitrary element of `G`.
+-/
+structure left_invariant_derivation extends derivation 𝕜 C^∞⟮I, G; 𝕜⟯ C^∞⟮I, G; 𝕜⟯ :=
+(left_invariant'' : ∀ g, 𝒅ₕ(smooth_left_mul_1 I g) (derivation.eval_at 1 to_derivation) =
```

Comment:
I think it's worth seeing if we can simplify this definition even further by filling in missing auxiliary definitions.

Given that we don't yet have a proper API for representations of group, I think we could try something like the following. Define the pull back of a derivation by `(g : G)`:
```lean
-- Or at least a raw function `→` with a `TODO` about promoting to a linear map.
def left_mul_pull_back :
  derivation 𝕜 C^∞⟮I, G; 𝕜⟯ C^∞⟮I, G; 𝕜⟯ →ₗ[𝕜] derivation 𝕜 C^∞⟮I, G; 𝕜⟯ C^∞⟮I, G; 𝕜⟯ :=
sorry
```
and then use it to define:
```lean
-- Or at least `has_scalar` with a `TODO` about promoting to a `mul_action`.
instance : mul_action G (derivation 𝕜 C^∞⟮I, G; 𝕜⟯ C^∞⟮I, G; 𝕜⟯) :=
{ smul     := λ g, left_mul_pull_back I g,
  one_smul := sorry,
  mul_smul := sorry, }
```
We should then be able to express `left_invariant''` as just:
```lean
(left_invariant'' : ∀ (g : G), g • to_derivation = to_derivation)
```

Reply:
Ok cool, I really like this idea! Do you have an idea of how this can be refactored when group representations will have proper APIs? (Just to see if this will need to be completely rewritten or not)

Reply:
Don't you get into trouble with diamonds with that action when `G = \kk`?

Reply:
Although more thought is needed for how to make that work for two-argument functions

Reply:
Good point about the diamond when `G = 𝕜`.

I think a good way to deal with this in this instance may be to generalise what I suggested to any self-diffeomorphism of a differentiable manifold, and not just left multiplication in a Lie group.

More precisely, instead of `left_mul_push_forward` above we could define (for any differentiable manifold `M`) something like:
```lean
def push_forward_of_diffeo (φ : M ≃ₘ^n⟮I, I'⟯ M) :
  derivation 𝕜 C^∞⟮I, M; 𝕜⟯ C^∞⟮I, M; 𝕜⟯ →ₗ[𝕜] derivation 𝕜 C^∞⟮I, M; 𝕜⟯ C^∞⟮I, M; 𝕜⟯ :=
sorry
```
and then define:
```lean
instance : mul_action (M ≃ₘ^n⟮I, I'⟯ M) (derivation 𝕜 C^∞⟮I, M; 𝕜⟯ C^∞⟮I, M; 𝕜⟯) :=
{ smul     := λ φ X, push_forward_of_diffeo I g X,
  one_smul := sorry,
  mul_smul := sorry, }
```
and finally define:
```lean
instance : has_coe G (G ≃ₘ^∞⟮I, I⟯ G) :=
⟨λ g,
{ to_fun                   := smooth_left_mul I g,
  inv_fun                  := smooth_left_mul I g⁻¹,
  left_inv                 := sorry,
  right_inv                := sorry,
  times_cont_mdiff_to_fun  := sorry,
  times_cont_mdiff_inv_fun := sorry, }⟩
```
which should allow us to avoid the diamond issue and write:
```lean
(left_invariant'' : ∀ (g : G), (g : G ≃ₘ^∞⟮I, I⟯ G) • to_derivation = to_derivation)
```

Reply:
> I feel like mathlib needs a type alias meaning "act on the domain in •", then the instances can be something like `has_scalar (on_domain R) (A \to B) := { smul := λ r f a, f ((r : R) • a)`

I agree about this more general point. We'll need to address it sooner or late, but not here I think.

Reply:
Unfortunately here left multiplication is not a diffeo. I wrote a counterexample in a reply to another comment of yours above.

Reply:
Oh I see, you mean [this comment](https://github.com/leanprover-community/mathlib/pull/8108#discussion_r661833799)

Hmm, is it worth going to extra trouble to support Lie monoids? Are there applications?

Reply:
Well, Lie monoids are more common than one might think. You can find many examples by taking submonoids of Lie groups that have borders, as [0,1], [0, +\infty) or many submonoids of Lie groups of matrices (often defined through inequalities). You can find some literature on Lie monoids here: https://newbedev.com/lie-groups-vs-lie-monoids. I don't think they are absolutely necessary, but they do exist in maths, there is literature on them, and there are some applications even if I don't need them for what I need to do later. I decided to implement them also because given that we were forced to implement Lie groups as manifolds with corners for technical reasons, this was also a nice mathematical justification to having borders involved in the theory.

I don't entirely agree with the fact that Lie monoids are causing "extra trouble" here. The way I see it is that you are proposing a more convenient formulation of left invariance for Lie groups, that I very much like, but that I would develop for Lie groups only: here we need to add other code more than replacing the existing one and I don't see why not developing what you say later in the file: to the final user does it matter if your formulation of the theory is incorporated in the definition or it is made available as a lemma? I feel the current definition is natural and having Lie monoids never was a problem for it.

Reply:
In case you don't have very strong reasons to incorporate what you say in the definition and you agree to develop your theory in the rest of the file I would rather start merging this PR and open a new one to implement your theory if this is ok with you, both because this PR is big already and because it would be better for reviewers I feel.

Reply:
Thanks for this, you've convinced me that it's worth supporting Lie monoids.

The main thing I wanted was the ability to state left-invariance in terms of a `mul_action` on `derivation 𝕜 C^∞⟮I, M; 𝕜⟯ C^∞⟮I, M; 𝕜⟯` but if necessary this could be independent work that establishes this only for Lie groups.

I'll come back and review again this evening after I finish some other bits of work.

Reply:
Great! No worries I guess we need to wait anyways for Eric's answer to the above comment (even though usually he answers very quickly)!

Reply:
OK let's stick with this definition and one day build the simpler version for Lie groups. Thanks for humouring me.